docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,500 @@
+implement a stateless Ceph-CSI plugin. It also details a naming scheme for
+volumes provisioned against a Ceph cluster and the resultant *volume_id* for
+the same that is shared with the CO system, enabling state to be carried by the
+*volume_id* where required.
+
+As of this writing, the plugin uses a config map to store *volume_name* to
+*volume_id* mapping, and also stores related Ceph cluster configuration
+information (e.g MON list, pool information) within the same tuple, to be able
+to find and operate on the provisioned volume. The implementation also depends
+on secrets from the StorageClass to perform NS related operations.
+
+The existing strategy enables a single CSI instance to operate across any Ceph
+cluster that is fully defined by the StorageClass, but brings in the burden of
+storing and managing provisioned volumes by the plugin.
+
+The proposal laid out here is to try and retain the ability for a single CSI
+instance to operate across any Ceph cluster, but remove the storage and
+management of provisioned CSI volumes by the plugin, or in short make the
+Ceph-CSI plugin stateless. Further, if this is achieved, it is also possible to
+run multiple replicas of the CS instance, to aid with load balancing of
+requests, than the current one instance limitation due to the existing config
+map.
+
+## Proposed solution details
+
+### Summary
+
+- Create a config map per Ceph cluster called `ceph-cluster-<cluster-fsid>` and
+  a pair of secrets to go along with the same cluster called
+  `ceph-cluster-<cluster-fsid>-provisioner-secret` and
+  `ceph-cluster-<cluster-fsid>-publish-secret`
+- Allow the CS and NS namespaces access to the above config maps and secrets,
+  in addition to the namespace that is responsible for creating and
+  maintaining it
+  - **Needs Investigation** Unsure if secrets and config maps can be
+    shared (via RBAC?) across namespaces or service accounts. This maybe
+    needed if we need operator managed Ceph instances to be able to pass
+    cluster configuration information to the CSI instance, and are in
+    different namespaces for e.g.
+- Details about the Ceph cluster are maintained in
+  `ceph-cluster-<cluster-fsid>`. Details include cluster-id, MONs, pools, and
+  other relevant data. Also, these and the corresponding secrets are updated
+  when they change, to reflect current values
+- The CSI plugin pod adds these config maps and secrets as volumes to its pod
+  specification, thus ensuring auto-refresh of content when these are updated
+  by the CO environment
+- VolumeCreate requests come against a targeted Ceph cluster-id and pool,
+  enabling the CSI plugin to locate the appropriate configuration information
+  for the cluster and its secrets to act on the request
+  - **NOTE:** The cluster-id and pool information is specified in the
+    StorageClass for example
+  - **NOTE:** In the **future**, if it is desired that the CSI plugin CS
```

Comment:
In the existing rook storage class (flex), we require the cluster and pool to be specified in the storage class. I haven't heard a scenario where that didn't work. In general, I wouldn't think the CS would have an opinion about this.

Reply:
The CS does not have any opinion on this, which is correct.

Specifying the cluster and pool from the storage class again will work just fine, and initial implementation would actully expect the same.

The proposal leaves the ability to not do so, and let the CSI driver choose which cluster and pool to use for the requested provisioning, if this information is missing. My reasons are two fold, but not **substantiated** enough,
- Simplify the StorageClass in such a manner, that just like asking for gp2 instances from AWS, the StorageClass asks for image type X from Ceph, and does not bother with where it comes from. Makes application pod requesting storage from a more generically named StorageClass possible.
- Topology and how that impacts where the storage comes from

So, just to reiterate, this is just a possible **future** direction that is enabled, how we want to reach there and if it is a reality, still needs data to conclude on the strategy.

Reply:
This is removed from the text. Retaining pool/fsname and clusterid to be specified in the storage class.